### PR TITLE
[BUG FIX] Add tests for adding multiple sensors and envs.

### PR DESCRIPTION
**Pull request by Milotrince**
## Description
Remove some code duplication, add at least two of each kind of sensors in tests and add fixture for 0 and 2 envs for coverage.

## Related Issue
Resolves https://github.com/Genesis-Embodied-AI/Genesis/issues/1722

## Motivation and Context
Bug :(

## How Has This Been / Can This Be Tested?
`pytest tests/test_sensors.py`

## Checklist:

- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [ ] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [x] I tested my changes and added instructions on how to test it for reviewers.

### CODE REVIEW

File: genesis/utils/ring_buffer.py [hunk 1]
```diff
@@ -67,7 +67,8 @@ def at(self, idx: int | torch.Tensor) -> torch.Tensor:
         rel_idx = (self._idx_ptr.value - idx) % self.N
         if isinstance(idx, torch.Tensor):
             batch_size = len(idx) if idx.ndim > 0 else 1
-            return self.buffer[rel_idx, :batch_size]
+            batch_indices = torch.arange(batch_size, device=gs.device, dtype=gs.tc_int)
+            return self.buffer[rel_idx, batch_indices]
```
Comment (Milotrince) [inline genesis/utils/ring_buffer.py]: I need to figure this out still.. It works when i do arange.. When using :batch_size and reshapes or squeeze(1) or having ifs i could not get it to work for both batched/unbatched case. I was tired when I tried last night though. Help is appreciated 🥲

File: genesis/utils/ring_buffer.py [hunk 2]
```diff
@@ -67,7 +67,8 @@ def at(self, idx: int | torch.Tensor) -> torch.Tensor:
         rel_idx = (self._idx_ptr.value - idx) % self.N
         if isinstance(idx, torch.Tensor):
             batch_size = len(idx) if idx.ndim > 0 else 1
-            return self.buffer[rel_idx, :batch_size]
+            # using [tensor_indices, slice] results in an extra dimension
+            return self.buffer[rel_idx, :batch_size][:, 0]
```
Comment (Milotrince) [inline genesis/utils/ring_buffer.py]: I think there is still something funky going on with the indexing and/or delay handling. I will try to add another test tomorrow.

File: genesis/sensors/contact_force.py [hunk 3]
```diff
@@ -127,9 +123,12 @@ def _update_shared_ground_truth_cache(
         cls, shared_metadata: ContactSensorMetadata, shared_ground_truth_cache: torch.Tensor
     ):
         all_contacts = shared_metadata.solver.collider.get_contacts(as_tensor=True, to_torch=True)
-        contact_links = torch.cat([all_contacts["link_a"], all_contacts["link_b"]], dim=-1)
-        is_contact = (contact_links.unsqueeze(-2) == shared_metadata.expanded_links_idx).any(dim=-1)
-        shared_ground_truth_cache.copy_(is_contact)
+        if all_contacts["link_a"].numel() == 0:
```
Comment (duburcqa) [inline genesis/sensors/contact_force.py:128]: DELETE COMMENT (I was wrong)

File: genesis/sensors/contact_force.py [hunk 4]
```diff
@@ -127,9 +123,12 @@ def _update_shared_ground_truth_cache(
...
+            shared_ground_truth_cache.fill_(False)
+        else:
+            contact_links = torch.cat([all_contacts["link_a"], all_contacts["link_b"]], dim=-1)
+            is_contact = (contact_links[..., -1, None] == shared_metadata.expanded_links_idx).any(dim=-1)
+            shared_ground_truth_cache.copy_(is_contact)
```
Comment (duburcqa) [inline genesis/sensors/contact_force.py]: Concatenating is probably less efficient that loop over the two arrays separately I think.

File: genesis/sensors/contact_force.py (hunk 3 above)
Comment (duburcqa) [reply genesis/sensors/contact_force.py:128]: Just do `is_contact = any(torch.isin(links_idx, shared_metadata.expanded_links_idx).any(dim=0) for links_idx in (all_contacts["link_a"], all_contacts["link_b"]))`

File: genesis/utils/ring_buffer.py (hunk 2 above)
Comment (duburcqa) [reply genesis/utils/ring_buffer.py]: I let you in the wrong direction here sorry. Now I understand what you are trying to achieve. In this specific case, you MUST use a tensor for the second index. But it is not an issue here, since you are already doing fancy indexing in the first place so allocating memory for the extracted buffer is expected... 

If you pass two tensors of the same length `indices_1, indices_2`, it will be interpreted as a tuple fancy index `(indices_1, indices_2)`, which will return a 1D vector (whatever the dimensionality of the input tensor) that corresponds to:
```
out[0] = in[indices_1[0], indices_2[0]]
out[1] = in[indices_1[1], indices_2[1]]
[...]
out[N] = in[indices_1[N], indices_2[N]]
```

If you mix a single tensor index with slices, it is completely different. It is a same as extracted sub-matrices that so stack by together, followed (or preceded, it is the same) by slicing on the other dimensions:
```
out = torch.stack([in[indices_1[i]] for i in indices_1], dim=0)[:, :N]
# Or equivalently
out[0, :] = in[indices_1[0], :N]
out[1, :] = in[indices_1[1], :N]
[...]
out[N, :] = in[indices_1[1], :N]
```

Not much in common!

Comment (duburcqa) [reply genesis/utils/ring_buffer.py]: You should update the doc to clarify what this method is supposed to do when a tensor index is specified.

By the way, the doc could be improved "Index of the element to get, where 0 is the latest element, 1 is the second latest, etc." --> "Index of the element to get, ordered from most recent to oldest [...]"

**Review by duburcqa — APPROVED** [review]

File: genesis/utils/ring_buffer.py (hunk 2 above)
Comment (duburcqa) [reply genesis/utils/ring_buffer.py]: Fixed!

File: genesis/sensors/contact_force.py (hunk 3 above)
Comment (duburcqa) [reply genesis/sensors/contact_force.py:128]: fixed!

File: genesis/sensors/contact_force.py (hunk 4 above)
Comment (duburcqa) [reply genesis/sensors/contact_force.py]: fixed!

**Review by duburcqa — DISMISSED** [review]

**Review by duburcqa — APPROVED** [review]
